logfield.go

```diff
@@ -0,0 +1,136 @@
+package opentracing
+
+import "math"
+
+const (
+	stringType fieldType = iota
+	boolType
+	intType
+	int64Type
+	float64Type
+	errorType
+	objectType
+	deferredObjectType
+)
+
+// LogField instances are constructed via LogBool, LogString, and so on.
+// Tracing implementations may then handle them via the LogField.Process
+// method.
+type LogField struct {
+	key          string
+	valType      valType
+	numericVal   int64
+	stringVal    string
+	interfaceVal interface{}
+}
+
+func LogString(key, val string) LogField {
+	return LogField{
+		key:       key,
+		valType:   stringType,
+		stringVal: val,
+	}
+}
+
+func LogBool(key string, val bool) LogField {
+	var numericVal int64
+	if val {
+		numericVal = 1
+	}
+	return LogField{
+		key:        key,
+		valType:    boolType,
+		numericVal: numericVal,
+	}
+}
+
+func LogInt(key string, val int) LogField {
+	var numericVal int64 = int64(val)
+	return LogField{
+		key:        key,
+		valType:    intType,
+		numericVal: numericVal,
+	}
+}
+
+func LogInt64(key string, val int64) LogField {
+	return LogField{
+		key:        key,
+		valType:    int64Type,
+		numericVal: val,
+	}
+}
+
+func LogFloat64(key string, val float64) LogField {
+	return LogField{
+		key:        key,
+		valType:    float64Type,
+		numericVal: int64(math.Float64bits(val)),
+	}
+}
+
+// REVIEWERS: etc etc for other numeric types if we like this direction
+
+func LogError(err error) LogField {
+	return LogField{
+		key:          "error",
+		valType:      errorType,
+		interfaceVal: err,
+	}
+}
+
+func LogObject(key string, obj interface{}) LogField {
+	return LogField{
+		key:          key,
+		valType:      objectType,
+		interfaceVal: obj,
+	}
+}
+
+type DeferredObjectGenerator func() interface{}
+
+func LogDeferredObject(key string, generator DeferredObjectGenerator) LogField {
+	return LogField{
+		key:          key,
+		valType:      deferredObjectType,
+		interfaceVal: generator,
+	}
+}
+
+// LogFieldProcessor allows access to the contents of a LogField (via a call to
+// LogField.Process).
+//
+// Tracer implementations typically provide an implementation of
+// LogFieldProcessor; OpenTracing callers should not need to concern themselves
+// with it.
+type LogFieldProcessor interface {
```

Comment:
Actually not really "helper" at the moment, as there is no way to get (otherwise unexported) state out of LogFields without this.

I don't like the naming at the moment either. Not sure if I like Visitor much more, but if we agree on the basic shape of the API I'm happy to take a pass at better names throughout.